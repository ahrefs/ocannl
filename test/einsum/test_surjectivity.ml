open Base
open Ocannl
module IDX = Train.IDX
module CDSL = Train.CDSL
open Nn_blocks.DSL_modules

module type Backend = Ir.Backend_intf.Backend

let test_diagonal_tensor () =
  Stdio.printf "\nTesting diagonal tensor initialization:\n";
  Tensor.unsafe_reinitialize ();
  let ctx = Context.auto () in
  (* Create a diagonal tensor using einsum: i->ii *)
  let input = TDSL.range 5 in
  let%op diagonal = input ++ "i=>ii" in

  (* Ensure the diagonal tensor is hosted *)
  Train.set_hosted diagonal.value;
  ignore (Train.forward_once ctx diagonal);

  (* Print the diagonal tensor *)
  Train.printf ~here:[%here] ~with_code:false ~with_grad:false diagonal;
  Stdio.printf "\n"

let test_sparse_fixed_index () =
  Stdio.printf "\nTesting sparse assignment with fixed index:\n";
  Tensor.unsafe_reinitialize ();
  let ctx = Context.auto () in
  (* Create a sparse tensor using fixed indices: i->i0j *)
  let input = TDSL.range 4 in
  let%op sparse = input ++ "i=>i0" in

  Train.set_hosted sparse.value;
  ignore (Train.forward_once ctx sparse);

  Train.printf ~here:[%here] ~with_code:false ~with_grad:false sparse;
  Stdio.printf "\n"

let test_multi_sparse () =
  Stdio.printf "\nTesting multiple sparse axes:\n";
  Tensor.unsafe_reinitialize ();
  let ctx = Context.auto () in
  (* Create tensor with multiple dims and test sparse assignment *)
  let input = TDSL.range_of_shape ~output_dims:[ 3; 4 ] () in
  let%op result = input ++ "ij=>i1j" in

  Train.set_hosted result.value;
  ignore (Train.forward_once ctx result);

  Train.printf ~here:[%here] ~with_code:false ~with_grad:false result;
  Stdio.printf "\n"

let _test_stride_gap () =
  Stdio.printf "\nTesting stride gap:\n";
  Tensor.unsafe_reinitialize ();
  let ctx = Context.auto () in
  (* Create tensor with multiple dims and test sparse assignment *)
  let input = TDSL.range_of_shape ~output_dims:[ 2; 5 ] () in
  let%op result = input ++ "ij=>i+3*j" in

  Train.set_hosted result.value;
  ignore (Train.forward_once ctx result);

  Train.printf ~here:[%here] ~with_code:false ~with_grad:false result;
  Stdio.printf "\n"

let () =
  test_diagonal_tensor ();
  test_sparse_fixed_index ();
  test_multi_sparse ();
  (* FIXME(#354): Projections inference for convolution-style expressions not implemented yet. *)
  (* test_stride_gap (); *)
  Stdio.printf "All surjectivity tests completed.\n"
