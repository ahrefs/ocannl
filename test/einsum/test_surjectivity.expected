Retrieving commandline, environment, or config file variable ocannl_log_level
Found 0, in the config file

Testing diagonal tensor initialization:
HERE: test/einsum/test_surjectivity.ml:23:21
┌──────────────────────────────────────┐
│[1]: =>_diagonal shape 0:6,1:6        │
│┌──────┬─────────────────────────────┐│
││      │axis 1                       ││
│├──────┼─────────────────────────────┤│
││axis 0│ 0.00  0.00  ...  0.00  0.00 ││
││      │ 0.00  1.00  ...  0.00  0.00 ││
││      │ ...   ...   ...  ...   ...  ││
││      │ 0.00  0.00  ...  4.00  0.00 ││
││      │ 0.00  0.00  ...  0.00  5.00 ││
│└──────┴─────────────────────────────┘│
└──────────────────────────────────────┘


Testing sparse assignment with fixed index:
HERE: test/einsum/test_surjectivity.ml:37:21
┌─────────────────────────────┐
│[1]: =>_sparse shape 0:5,1:1 │
│┌──────┬──────┐              │
││      │axis 1│              │
│├──────┼──────┤              │
││axis 0│ 0.00 │              │
││      │ 1.00 │              │
││      │ 2.00 │              │
││      │ 3.00 │              │
││      │ 4.00 │              │
│└──────┴──────┘              │
└─────────────────────────────┘


Testing multiple sparse axes:
HERE: test/einsum/test_surjectivity.ml:51:21
┌───────────────────────────────────────┐
│[1]: =>_result shape 0:3,1:2,2:4       │
│┌──────┬──────────────────────────────┐│
││      │axis 2                        ││
│├──────┼──────────────────────────────┤│
││0 @ 0 │ 0.00  0.00  0.00  0.00       ││
││axis 1│ 0.00  1.00  2.00  3.00       ││
│├──────┼──────────────────────────────┤│
││1 @ 0 │ 0.00  0.00  0.00  0.00       ││
││axis 1│ 4.00  5.00  6.00  7.00       ││
│├──────┼──────────────────────────────┤│
││2 @ 0 │ 0.00  0.00  0.00     0.00    ││
││axis 1│ 8.00  9.00  1.00e+1  1.10e+1 ││
│└──────┴──────────────────────────────┘│
└───────────────────────────────────────┘

All surjectivity tests completed.
